numpy/core/src/multiarray/descriptor.c

```diff
@@ -3166,10 +3168,25 @@ arraydescr_struct_str(PyArray_Descr *dtype, int includealignflag)
      */
     if (!(includealignflag && (dtype->flags&NPY_ALIGNED_STRUCT)) &&
                         is_dtype_struct_simple_unaligned_layout(dtype)) {
-        return arraydescr_struct_list_str(dtype);
+        sub = arraydescr_struct_list_str(dtype);
     }
     else {
-        return arraydescr_struct_dict_str(dtype, includealignflag);
+        sub = arraydescr_struct_dict_str(dtype, includealignflag);
+    }
+    
+    /* If the data type has a nonstandard (subclassed) type, show it */
+    if(dtype->type_num == NPY_VOID && dtype->typeobj != &PyVoidArrType_Type) {
+        PyObject *typestr = PyUString_FromString(dtype->typeobj->tp_name);
+
+        PyObject *ret = PyUString_FromString("(");
+        PyUString_ConcatAndDel(&ret, typestr);
+        PyUString_ConcatAndDel(&ret, PyUString_FromString(", "));
+        PyUString_ConcatAndDel(&ret, sub);
+        PyUString_ConcatAndDel(&ret, PyUString_FromString(")"));
+        return ret;
```

Comment:
I don't think so, since PyUString_ConcatAnd*Del* decrefs it.

I largely copied the code from arraydescr_subarray_str just below. That code does no decrefs, nor does it check the return values.